grudge/trace_pair.py

```diff
@@ -276,57 +284,109 @@ class _RankBoundaryCommunication:
     base_tag = 1273
 
     def __init__(self, dcoll: DiscretizationCollection,
-                 remote_rank, vol_field, tag=None):
+                 remote_rank, array_container, tag=None):
         self.tag = self.base_tag
         if tag is not None:
             self.tag += tag
 
         self.dcoll = dcoll
-        self.array_context = vol_field.array_context
+        self.array_context = get_container_context_recursively(array_container)
         self.remote_btag = BTAG_PARTITION(remote_rank)
         self.bdry_discr = dcoll.discr_from_dd(self.remote_btag)
 
         from grudge.op import project
 
-        self.local_dof_array = project(dcoll, "vol", self.remote_btag, vol_field)
+        self.local_array_ctr = \
+            project(dcoll, "vol", self.remote_btag, array_container)
+
+        # NOTE: Want to remember ordering of entries so the ordering of the keys
+        # corresponds to offset calculations for unflattening the exchanged array
+        container_data = OrderedDict()
+        flattened_data = []
+        for key, value in serialize_container(self.local_array_ctr):
+            flattened_values = rec_map_dof_array_container(
+                partial(flatten_to_numpy, self.array_context), value
+            )
+            container_data[key] = flattened_values
+            # NOTE: Addresses object arrays inside array containers. Need to special
+            # case them here to make sure we can convert it into a flat numpy array
+            # that mpi4py can actually send/recieve
+            if (isinstance(flattened_values, np.ndarray)
+                    and flattened_values.dtype == "O"):
+                flattened_values = np.concatenate(flattened_values)
+            flattened_data.append(flattened_values)
```

Comment:
The reason for this `if`-statement here is that I'm trying to avoid _all_ sub array containers (in particular, object numpy arrays). `meshmode.dof_array.flatten_to_numpy` (and similar functions) appear to only be implemented with flattening group-wise in mind. As a result, I have to add these special cases (same in `finish()` when converting _back_ to an array container).